### PR TITLE
aws: Create cluster without DNS or Gossip

### PR DESCRIPTION
**Pull request by hakman**
Follow-up PR to https://github.com/kubernetes/kops/pull/14440.

/hold for #14440 to merge first

/cc @justinsb @johngmyers @olemarkus

### CODE REVIEW

**Comment by hakman** [comment]
Looks like e2e tests pass without issues using `--dns=none`.

**Comment by johngmyers** [comment]
We should make presubmits for dnsnone and apiserver-nodes-dnsnone. (Unless the combination of apiserver-nodes and dnsnone doesn't make sense, in which case we should catch that in API validation.)

**Comment by hakman** [comment]
> We should make a presubmit for apiserver-nodes-dns-none. (Unless the combination of apiserver-nodes and dnsnone doesn't make sense, in which case we should catch that in API validation.)

I added an extra check to keep the `kops-controller` target group limited to master asgs. The apiserver related target groups are also added to apiserver asgs.

I also added an e2e template for APIServer nodes with LB and without DNS.

**Comment by hakman** [comment]
In the end, seems that APIServer nodes with LB and without DNS will not work for now, due to the lack of access to etcd, which would require additional target groups (if at all possible).

**Comment by olemarkus** [comment]
You'd need an ELB per cluster, probably (since different etcd clusters can have different sizes/IGs)

**Comment by hakman** [comment]
> You'd need an ELB per cluster, probably (since different etcd clusters can have different sizes/IGs)

Why not just different target groups added to the apiserver load balancer, each with his own listening port?

**Comment by olemarkus** [comment]
Yeah that could be possible. I don't know what would be the least complex there.

**Comment by hakman** [comment]
> Yeah that could be possible. I don't know what would be the least complex there.

It's also about the cost. One more LB would mean at least an extra USD 30.

**Comment by johngmyers** [comment]
Does the combination make sense? Would a cluster that is big enough to need apiserver nodes want to run without DNS?

**Comment by hakman** [comment]
> Does the combination make sense? Would a cluster that is big enough to need apiserver nodes want to run without DNS?

Not really. I think as is it makes sense to get started. Would make that part a little easier by removing any need for a DNS zone.

**Comment by johngmyers** [comment]
Someone getting started wouldn't need apiserver node. I think we should just add API validation to prohibit this combination and move on.

**Comment by hakman** [comment]
> Someone getting started wouldn't need apiserver node. I think we should just add API validation to prohibit this combination and move on.

That's what I meant. Already added.

**Review by olemarkus — COMMENTED** [review]
/lgtm

**Comment by olemarkus** [comment]
Add release notes maybe?

**Comment by hakman** [comment]
> Add release notes maybe?

Will add as a separate PR.

File: pkg/model/awsmodel/api_loadbalancer.go [hunk 1]
```diff
@@ -521,6 +557,33 @@ func (b *APILoadBalancerBuilder) Build(c *fi.ModelBuilderContext) error {
 		}
 	}
 
+	// Allow kops-controller to the master instances from the ELB
+	if b.Cluster.UsesNoneDNS() && b.APILoadBalancerClass() == kops.LoadBalancerClassNetwork {
```
Comment (johngmyers) [inline pkg/model/awsmodel/api_loadbalancer.go:561]: Should we prohibit NoneDNS with Classic API loadbalancers?

**Review by johngmyers — APPROVED** [review]
/approve
/hold for question in review

File: pkg/model/awsmodel/api_loadbalancer.go (hunk 1 above)
Comment (hakman) [reply pkg/model/awsmodel/api_loadbalancer.go:561]: Sure, done.

**Pull request merged**